### ISSUE TITLE
creating a lot of new nodes during pre-disovery causes conserver a long time to acquire locks.

### ISSUE THREAD

**Issue by ralphbellofatto** [xcat2/xcat-core#1801]
We Are attempting to use MTMS based discovery so we created a stanza file containing 80+ new nodes.

When we do this and then startup conserver, we notice that it takes a long for the console to acquire their lock before we can use them.

we have setup our site configuration with "consoleondemand=no" so we are logging all of the console traffic all of the time.

I have attached the stanza file with the new node definitions here:
[newnodes.stanza.txt](https://github.com/xcat2/xcat-core/files/451867/newnodes.stanza.txt)

when attempting to try to find out why all the console would show "Acquiring lock.... for a really long time we instrumented the code in:
/opt/xcat/share/xcat/cons/ipmi
to try to tell what is ho long the it takes to acquire the lock..

```
sub get_lock {
    unless (flock(LOCKHANDLE, LOCK_EX | LOCK_NB)) {
        $| = 1;
        print "$ARGV[0] Acquiring startup lock...";
        ($seconds, $microseconds) = gettimeofday;  $starttime=$seconds+($microseconds/1000000);

        flock(LOCKHANDLE, LOCK_EX) or die "Fatal Error securing startup lock";
        ($seconds, $microseconds) = gettimeofday;  $endtime=$seconds+($microseconds/1000000);
        $deltatime = $endtime-$starttime;
        print "done $deltatime MS\n";
    }
    truncate(LOCKHANDLE, 0);
    print LOCKHANDLE $$ . "\n";
}
```

When examining the code in the /var/log/consoles/*

we see acquire lock times between 30 and 4000 seconds.

```
[Thu Sep  1 23:45:16 2016]c460c043 Acquiring startup lock...done 33.6890850067139 MS
[Fri Sep  2 00:51:30 2016]c460c090 Acquiring startup lock...done 4661.84337496758 MS
```

And the problem persists and never stabilizes, making trying to use the system.

Here are the times to acquire the locks for nodes that have bmc's that we can connect to.

```
c460c001 Acquiring startup lock...done 265.741617918015 MS
c460c002 Acquiring startup lock...done 1489.7107770443 MS
c460c003 Acquiring startup lock...done 2173.9434659481 MS
c460c004 Acquiring startup lock...done 223.871228933334 MS
c460c005 Acquiring startup lock...done 95.7902700901031 MS
c460c007 Acquiring startup lock...done 280.591845035553 MS
```

So over 2000 seconds 30 minutes to acquire a new lock to observe a node we are bringing up.

This makes using conondemand=no during site discovery a bit problematic...

Since the CORAL contract requires that console logs be captured all the time, this creates a problem whenever BMC's can't be contacted, either during bringup or with switch outage.

**Comment by whowutwut** [comment]
I haven't kept up with the changes that @jjohnson42  has been making into confluent.   But I do think that we should get confluent-deps re-built on Power operating systems and really have the abililty to replace conserver with confluent  for console management..  We can do a performance test for confluent and see whether we even have this issue at all, if not, maybe that's the solution we have moving forward in xCAT.   Thoughts?

**Comment by ralphbellofatto** [comment]
I don't think this is a confluent vs xcat problem.  I think the problem has to do with the fact that the getbmc getipmicons call is done under lock a file lock in /tmp/xcat/consolelock and when a bmc is unreachable, there is a 10 second or so timeout.

This is done in in the /opt/xcat/share/xcat/cons/ipmi file.

This results in a lot of consoles waiting for locks.

I have found a temporary way around this is to hand edit the conserver.cf file and add ondemand to each device that is not responding.

there is a console on demand setting that can be attached to the lsdef for the node, but it does not work reliably in changing this value in the conserver.cf file once it is defined.  

Which is the subject for another bug..

**Comment by gurevichmark** [comment]
I will take a look at what can be done.

**Comment by gurevichmark** [comment]
No sure how to fix this issue yet. Moving to next release.

**Comment by ralphbellofatto** [comment]
The reason this coarse grained lock needs to be examined and then the way the lock is done so that it is more fine grained.

I suspect the reason for the lock is to avoid having two consoles trying to get information from the same BMC at the same time.

However, the lock go across access to all the BMC's.    

Currently this means you cannot have console on demand as the default.  

It also means, that when you have an event when a lot of nodes have their physical power turned off either by pulling the plug or at the power distribution unit that someone has to remember to set consoleondemand=1 on those nodes otherwise, the overall console logs will stop working.

**Comment by immarvin** [comment]
hi @zet809 , would you please take a look at this issue?
thanks

**Comment by immarvin** [comment]
hi @chenglch , any update?

**Comment by whowutwut** [comment]
@ralphbellofatto  If you can verify and close out, please do, otherwise @tingtli  Can you assign FVT team to help verify and close out?

**Comment by ralphbellofatto** [comment]
will the fix for this be picked up if i do an *xCAT*??

**Comment by whowutwut** [comment]
yes, I think so

**Comment by tingtli** [comment]
@junxiawang will you please help verify this bug

**Comment by ralphbellofatto** [comment]
To test this, create a lot of new nodes with the following template:
```
testnodes008:
    objtype=node
    usercomment=[test]
    groups=all
    consoleondemand=0
    bmc=10.20.4.8
    bmcport=0
    cons=ipmi
    installnic=mac
    ip=10.21.4.8
    mgt=ipmi
    netboot=petitboot
    primarynic=mac

```



then do this (which will restart the conserver).
```
cat newnode.lsdef | chdef -z
date
date
makeconservercf
```

```
Tue Nov 22 09:09:14 EST 2016




and then check the console log...

[consoleup.log](/uploads/33e3412db410051c8d4eceb9edc75e64/consoleup.log)

Seems we have a range to get the initial lock from almost immediate to about 2 minutes, which for this number of nodes is an acceptable range for the number of nodes that this was reported for.


```
[root@c460mgt01 ralphbel]#  grep "Console up" -A3 /var/log/consoles/c460c* | grep "Nov 22 09" | tee /tmp/consoleup.log
/var/log/consoles/c460c001:[-- Console up -- Tue Nov 22 09:09:15 2016]
/var/log/consoles/c460c001-[Tue Nov 22 09:10:05 2016]Info: SOL payload already de-activated
/var/log/consoles/c460c001-[Tue Nov 22 09:10:05 2016][SOL Session operational.  Use ~? for help]
...
/var/log/consoles/c460c003:[-- Console up -- Tue Nov 22 09:09:15 2016]
/var/log/consoles/c460c003-[Tue Nov 22 09:09:15 2016]Info: SOL payload already de-activated
/var/log/consoles/c460c003-[Tue Nov 22 09:09:16 2016][SOL Session operational.  Use ~? for help]
...
/var/log/consoles/c460c036:[-- Console up -- Tue Nov 22 09:09:16 2016]
/var/log/consoles/c460c036-[Tue Nov 22 09:11:13 2016]Info: SOL payload already de-activated
/var/log/consoles/c460c036-[Tue Nov 22 09:11:14 2016][SOL Session operational.  Use ~? for help]
```

I wonder how this scales when we add even more nodes...

**Comment by chenglch** [comment]
Yes, this is a problem that conserver will maintain a process for each node and the process forked by conserver will send request to xcatd(this step need 3 processes) which also occupied a  lot of cpu and memory resources.  This is advantage of confluent which leverage the `eventlet` to avoid too many processes.  

We are trying to monitor the resource usage when restarting the conserver to see how many nodes should be managed by SN.

**Comment by ralphbellofatto** [comment]
ok, well, at least with 80 new nodes, the time it takes to get a lock is relatively acceptable to what it was.

What is the status of confluent, and do you know if it will be recommended to be deployed with CORAL?

**Comment by whowutwut** [comment]
Closing this out.  @ralphbellofatto If the BMC could directly redirect console output using syslog, would that be a good alternative for this?  If so, maybe we can open a new issue to track that..

**Comment by ralphbellofatto** [comment]
Getting a scale-able console log needs to be something that is needed for CORAL before they scale up to their 4k nodes.

One thing that will be very important in rsyslog is that the node has to be identified in the system by its lsf name.

Past attempts to have the BMC send stuff directly to rsyslog (not the console log, but other stuff), logged the information under the BMC's ip address, which makes it hard to use the data in rsyslog since there is a manual lookup step to match the ip address to a name.
